src/portfolio.py

```diff
@@ -0,0 +1,297 @@
+        for portfolio in self.portfolios:
+            if portfolio.rebalancing_date == rebalancing_date:
+                return portfolio.weights
+        return None
+
+    def get_weights_df(self) -> pd.DataFrame:
+        weights_dict = {}
+        for portfolio in self.portfolios:
+            weights_dict[portfolio.rebalancing_date] = portfolio.weights
+        return pd.DataFrame(weights_dict).T
+
+    def get_portfolio(self, rebalancing_date: str) -> Portfolio:
+        if rebalancing_date in self.get_rebalancing_dates():
+            idx = self.get_rebalancing_dates().index(rebalancing_date)
+            return self.portfolios[idx]
+        else:
+            raise ValueError(f'No portfolio found for rebalancing date {rebalancing_date}')
+
+    def has_previous_portfolio(self, rebalancing_date: str) -> bool:
+        dates = self.get_rebalancing_dates()
+        ans = False
+        if len(dates) > 0:
+            ans = dates[0] < rebalancing_date
+        return ans
+
+    def get_previous_portfolio(self, rebalancing_date: str) -> Portfolio:
+        if not self.has_previous_portfolio(rebalancing_date):
+            raise ValueError(f'No portfolio prior to rebalancing date {rebalancing_date}')
+        else:
+            yesterday = [x for x in self.get_rebalancing_dates() if x < rebalancing_date][-1]
+            return self.get_portfolio(yesterday)
+
+    def get_initial_portfolio(self, rebalancing_date: str) -> Portfolio:
+        if self.has_previous_portfolio(rebalancing_date = rebalancing_date):
+            initial_portfolio = self.get_previous_portfolio(rebalancing_date)
+        else:
+            initial_portfolio = Portfolio(rebalancing_date = None, weights = {})
+        return initial_portfolio
+
+    def __repr__(self):
+        return f'Strategy(portfolios={self.portfolios})'
+
+    def number_of_assets(self, th: float = 0.0001) -> pd.Series:
+        return self.get_weights_df().apply(lambda x: sum(np.abs(x) > th), axis = 1)
+
+    def turnover(self, return_series, rescale = True):
+
+        dates = self.get_rebalancing_dates()
+        to = {}
+        to[dates[0]] = 1
+        for rebalancing_date in dates[1:]:
+            previous_portfolio = self.get_previous_portfolio(rebalancing_date)
+            current_portfolio = self.get_portfolio(rebalancing_date)
+            to[rebalancing_date] = current_portfolio.turnover(portfolio = previous_portfolio,
+                                                              return_series = return_series,
+                                                              rescale = rescale)
+        return pd.Series(to)
+
+    def simulate(self,
+                 return_series = None,
+                 fc: float = 0,
+                 vc: float = 0,
+                 n_days_per_year: int = 252) -> pd.Series:
+
+        rebdates = self.get_rebalancing_dates()
+        ret_list = []
+        for rebdate in rebdates:
+            if rebdate < rebdates[-1]:

```

Comment:
Alternatively, this could be written in one line
```
end_date = rebdates[rebdates.index(rebdate) + 1] if rebdate < rebdates[-1] else return_series.index[-1]
```